distributed/scheduler.py

```diff
@@ -1635,6 +1637,37 @@ def get_dataset(self, stream, name=None, client=None):
         else:
             raise KeyError("Dataset '%s' not found" % name)
 
+    @gen.coroutine
+    def register_environments(self, stream=None, environments=None, workers=None):
+        if environments is None:
+            environments = self.environments
+        else:
+            for name in environments:
+                if name in self.environments:
+                    raise KeyError("Environment %s already exists" % name)
+
+        if workers is None:
+            workers = list(self.worker_info)
+
+        results = yield [self.rpc(addr=worker).register_environments(environments=environments)
+                         for worker in workers]
+
+        exceptions = {}
+        for worker, result in zip(workers, results):
+            for name, val in result.items():
+                if isinstance(val, bool):
+                    if name not in self.environments:
+                        self.environments[name] = environments[name]
```

Comment:
Perhaps do `self.environments.upate(environments)` early on?  This looks complex-ish.

For example, if we our Scheduler has no workers when this is run with then we don't hold onto the environment for future workers.